runtime/cudaq/platform/mqpu/remote/RemoteSimulatorQPU.cpp

```diff
@@ -0,0 +1,93 @@
+#include "cudaq/platform/quantum_platform.h"
+
+namespace {
+using namespace mlir;
+
+// Remote QPU: delegating the execution to a remotely-hosted server, which can
+// reinstate the execution context and JIT-invoke the kernel.
+class RemoteSimulatorQPU : public cudaq::QPU {
+private:
+  std::string m_simName;
+  std::unordered_map<std::thread::id, cudaq::ExecutionContext *> m_contexts;
+  std::mutex m_contextMutex;
+  std::unique_ptr<MLIRContext> m_mlirContext;
+  std::unique_ptr<cudaq::RemoteRuntimeClient> m_client;
+
+public:
+  RemoteSimulatorQPU()
+      : QPU(), m_mlirContext(cudaq::initializeMLIR()),
+        m_client(cudaq::registry::get<cudaq::RemoteRuntimeClient>("rest")) {}
+
+  std::thread::id getExecutionThreadId() const {
+    return execution_queue->getExecutionThreadId();
+  }
+
+  virtual void setTargetBackend(const std::string &backend) override {
+    auto parts = cudaq::split(backend, ';');
+    if (parts.size() % 2 != 0)
+      throw std::invalid_argument("Unexpected backend configuration string. "
+                                  "Expecting a ';'-separated key-value pairs.");
+    for (std::size_t i = 0; i < parts.size(); i += 2) {
+      if (parts[i] == "url")
+        m_client->setConfig({{"url", parts[i + 1]}});
+      if (parts[i] == "simulator")
+        m_simName = parts[i + 1];
+    }
+  }
+
+  void enqueue(cudaq::QuantumTask &task) override {
+    cudaq::info("RemoteSimulatorQPU: Enqueue Task on QPU {}", qpu_id);
+    execution_queue->enqueue(task);
+  }
+
+  void launchKernel(const std::string &name, void (*kernelFunc)(void *),
+                    void *args, std::uint64_t voidStarSize,
+                    std::uint64_t resultOffset) override {
+    cudaq::info("RemoteSimulatorQPU: Launch kernel named '{}' remote QPU {} "
+                "(simulator = {})",
+                name, qpu_id, m_simName);
+
+    cudaq::ExecutionContext *executionContext = [&]() {
+      std::scoped_lock<std::mutex> lock(m_contextMutex);
+      const auto iter = m_contexts.find(std::this_thread::get_id());
+      if (iter == m_contexts.end())
+        throw std::runtime_error("Internal error: invalid execution context");
+      return iter->second;
+    }();
+    std::string errorMsg;
+    const bool requestOkay =
+        m_client->sendRequest(*m_mlirContext, *executionContext, m_simName,
+                              name, kernelFunc, args, voidStarSize, &errorMsg);
+    if (!requestOkay)
+      throw std::runtime_error("Failed to launch kernel. Error: " + errorMsg);
+  }
+
+  void setExecutionContext(cudaq::ExecutionContext *context) override {
+    cudaq::info("RemoteSimulatorQPU::setExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
+    m_contexts[std::this_thread::get_id()] = context;
+  }
+
+  void resetExecutionContext() override {
+    cudaq::info("RemoteSimulatorQPU::resetExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
```

Comment:
Just curious ... is there a reason `std::scoped_lock` was chosen instead of `std::unique_lock` (65, 81, and 87). My understanding is that `std::scoped_lock` is typically used when multiple locks are needed, but that doesn't seem to be the case here.

Reply:
I always think that `scoped_lock` is a simple scope-based lock whereas `unique_lock` has some ownership semantics.

I didn't notice until now that `scoped_lock` is listed in cppreference as "deadlock-avoiding RAII wrapper for _multiple_ mutexes". Looking at its implementation, it's the most efficient one for this simple use case (lock and unlock at ctor and dtor) vs. an extra variable required for `unique_lock`.